File: java/CircularQueue.java
// linear data structure that operates in a circular manner 
// overcomes limitation of fixed size in a simple array-based queue by reusing the space once
// the rear reaches the end of the array 

public class CircularQueue {
  private int[] arr;
  private int front, rear, capacity;

  public CircularQueue(int size) {
    capacity = size;
    arr = new int[capacity];
    front = rear = -1;
  }
  public boolean isEmpty() {
    return front == -1;
  }
  public boolean isFull() {
    return (rear + 1) % capcity == front;
  }
  public void enqueue(int item) {
    if(isFull()) {
      System.out.println("queue is full");
      return;
    }
    if(isEmpty()) {
      front = rear = 0;
    } else {
      rear = (rear + 1) % capacity;
    }
    arr[rear] = item;
    System.out.println(item + "enqueued to queue");
  }
  public int dequeue() {
    if(isEmpty()) {
      System.out.println("queue is empty");
      return -1;
    }
    int item = arr[front];
    if(front == rear) {
      front = rear = -1;
    } else {
      front = (front + 1) % capacity;
    }
    System.out.println(item + "dequeued from queue");
    return item;
  }
  public int peek() {
    if(isEmpty()) {
      System.out.println("queue is empty");
      return -1;
    }
    return arr[front];
  }
}

File: javascript/CircularQueue.js
// linear data structure that operates like a queue but with a fixed size 
// once queue is full, next element to be inserted overwrites oldest element

class CircularQueue {
  constructor(size) {
    this.queue = new Array(size);
    this.front = this.rear = -1;
    this.size = size;
  }
  isEmpty() {
    return this.front === -1;
  }
  isFull() {
    return (this.rear + 1) % this.size === this.front;
  }
  enqueue(item) {
    if(this.isFull()) {
      console.log("queue is full");
      return;
    }
    if(this.isEmpty()) {
      this.front = this.rear = 0;
    } else {
      this.rear = (this.rear + 1) % this.size;
    }
    this.queue[this.rear] = item;
  }
  dequeue() {
    if(this.isEmpty()) {
      console.log("queue is empty");
      return;
    }
    const item = this.queue[this.front];
    if(this.front === this.rear) {
      this.front = this.rear = -1;
    } else {
      this.front = (this.front + 1) % this.size;
    }
    return item;
  }

  printQueue() {
    if(this.isEmpty()) {
      console.log("queue is empty");
      return;
    }
    let i = this.front;
    while(i != this.rear) {
      console.log(this.queue[i]);
      i = (i + 1) % this.size;
    }
    console.log(this.queue[i]);
  }
}
